### PR TITLE
GAIL 4 Bark - Practical SS2020 

### PR DESCRIPTION
**Pull request by Brucknem**
Pull request for the Generative Adverserial Imitation Learning (GAIL) implementation.
The fork includes all work done as the project for the TUM practical course in the summer semester of 2020.
 
## Final report

To access the final report/jupyter notebook: `bazel run //docs/report:run`

***

## Running the examples/gail.py script to test our implementation

We have included pre-generated expert trajectories on the highway and merging blueprint. You can train a GAIL agent on these using the _examples/gail.py_ script. To start training a new GAIL agent, run `bazel run //examples:gail -- --mode=train`. To visualize a pretrained GAIL agent, run `bazel run //examples:gail -- --mode=visualize`.
Please have a look at the _examples/gail.py_ script for in depth explanation on how to use the expert trajectories for training and on how to use the pre-trained agents.

***

If you want to train a new SAC expert agent, use it to generate new expert trajectories and train a GAIL agent on these, perform the following:
1.	Train SAC agent: `bazel run //examples:tfa_generate -- --mode=train`
2.	Generate Expert trajectories: `bazel run //examples:tfa_generate -- --mode=generate`
	The output path is set as "GenerateExpertTrajectories::OutputDirectory"
3.	Set path of generated expert trajectories in examples/example_params/gail_params.json --> "ML::ExpertTrajectories::expert_path_dir"	
4.	Train GAIL agent: `bazel run //examples:gail -- --mode=train`
	The training logs and checkpoints are specified in "ML::GAILRunner::logdir" and "ML::GAILRunner::model_dir"
5.	Visualize GAIL agent: `bazel run //examples:gail -- --mode=visualize`
	You need to adjust the "ML::GAILRunner::logdir" and ""ML::GAILRunner::model_dir" as described in the _examples/gail.py_ script to use your own agent.
	
***

### Problems using the Interaction Dataset

Unfortunately, training does not completely work with the interaction dataset itself. Hence, the intermediate step of generating expert trajectories with a pre-trained SAC agent.

There are efforts to train a GAIL agent on the interaction dataset as described in [this development branch](https://github.com/GAIL-4-BARK/bark-ml/pull/73). We can train an agent for a huge number of episodes, but always end up in a state of mode collapse where the agent always generates the same action for any state. Thus, we decided not to merge these changes into the master branch.

***

### (Main) files added:

- _bark_ml/library_wrappers/lib_tf2rl/agents/gail_agent.py_
GAIL specific class

- _bark_ml/library_wrappers/lib_tf2rl/agents/tf2rl_agent.py_
Super class for TF2RL agents
- _bark_ml/library_wrappers/lib_tf2rl/compare_trajectories.py_
Evaluation functionalities used to compare our agent with the SAC expert
- _bark_ml/library_wrappers/lib_tf2rl/generate_expert_trajectories.py_
Script for generating expert trajectories from the Interaction Dataset
- _bark_ml/library_wrappers/lib_tf2rl/load_expert_trajectories.py_
Script for loading expert trajectories for GAIL
- _bark_ml/library_wrappers/lib_tf2rl/load_save_utils.py_
Helper functions for file import/export
- _bark_ml/library_wrappers/lib_tf2rl/normalization_utils.py_
Helper functions for normalization during training
- _bark_ml/library_wrappers/lib_tf2rl/runners/gail_runner.py_
GAIL specific runner class
- _bark_ml/library_wrappers/lib_tf2rl/runners/tf2rl_runner.py_
Super class for TF2RL runner_
- _bark_ml/library_wrappers/lib_tf2rl/tf2rl_wrapper.py_
Wraps the BARK runtime to match the expectations of TF2RL about the environment. The observation and action normalization also takes place here.
- _bark_ml/tests/py_library_tf2rl_tests/_
Folder containing the majority of our tests
- _docs/report/data/params/gail_params.json_
Special parameter file for the final jupyter notebook
- _docs/report/gail.ipynb_
Our final report/jupyter notebook – run by `bazel run //docs/report:run`
- _examples/example_params/gail_params.json_
Parameter file for GAIL – most notably, set path of expert trajectories 
- _examples/example_params/tfa_generate_params.json_
Special parameter file for generating expert trajectories from a pre-trained SAC agent – most notably, set path for export of expert trajectories 
- _examples/gail.py_
Entrance point for executing GAIL training, visualization and evaluation – run by 
`bazel run //examples:gail -- --mode=train`
- _examples/tfa_generate.py_
Entrance point for generating expert trajectories from a pre-trained SAC agent – run by `bazel run //examples:tfa_generate -- --mode=generate`
- _utils/dependencies.bzl_
Additional dependencies needed for bazel to build the runnable programs
- _utils/python/generate_launch_configuration.py_
Script for easy set up of launch configurations 
- _utils/tests/generate_launch_configuration_test.py_
Corresponding test

### (Main) files adjusted
- _bark_ml/library_wrappers/lib_tf_agents/runners/sac_runner.py_
Added the SACRunnerGenerator class for generating expert trajectories

- _bark_ml/observers/nearest_observer.hpp_
Added a flag to observe not normalized states of the agents
- _bark_ml/observers/nearest_state_observer.py_
Added a flag to observe not normalized states of the agents
- _utils/docker/installers/requirements.txt_
Python environment adjusted to our needs

***

## The implementation depends on
1. The forked [TF2RL](https://github.com/GAIL-4-BARK/tf2rl) repository for adjustments allowing for a better functioning with BARK and 
2. [Our own repository](https://github.com/GAIL-4-BARK/large_data_store) where pre-trained SAC and GAIL agents, both running on the merging and the highway blueprint, are stored as well as expert trajectories from the interaction dataset, from a pre-trained SAC agent on the merging blueprint, from a pre-trained SAC agent on the highway blueprint and from the Interaction Dataset

***

TF2RL does not work properly with the newest tensorflow version 2.3, hence ensuring version 2.2

### CODE REVIEW

File: bark_ml/environments/single_agent_runtime.py [hunk 1]
```diff
@@ -61,6 +61,8 @@ def step(self, action):
 
     # step and observe
     self._world.Step(self._step_time)
+    if eval_id not in self._world.agents:
+      return [0] * self._observer.observation_space.shape[0], -1, True, None
     observed_world = self._world.Observe([eval_id])[0]
```
Comment (Brucknem) [inline bark_ml/environments/single_agent_runtime.py:66]: Adding this change was necessary to train the SAC, PPO and GAIL agents on the highway and intersection blueprint. After the `Step()` function sometimes the eval agent is not valid anymore and thus the `Observe()` method raises an IndexError. 
We decided to fix it by returning a mock observation, negative reward, done flag set and no further environment info.

Comment (patrickhart) [reply bark_ml/environments/single_agent_runtime.py:66]: If the agent is not valid anymore, why is the episode not terminal?
Additionally, there is a flag: `params["World"]["remove_agents_out_of_map"] = True`

File: bark_ml/library_wrappers/lib_tf_agents/tfa_wrapper.py [hunk 2]
```diff
@@ -7,7 +7,9 @@
 from tf_agents.trajectories import time_step as ts
 
 tf.compat.v1.enable_v2_behavior()
-
+config = tf.compat.v1.ConfigProto()
+config.gpu_options.allow_growth = True
```
Comment (Brucknem) [inline bark_ml/library_wrappers/lib_tf_agents/tfa_wrapper.py:11]: This prevents that Tensorflow allocates 100% of the GPU memory. With this change multiple TFA agents can be trained simultaneously.

File: bark_ml/environments/single_agent_runtime.py (hunk 1 above)
Comment (Brucknem) [reply bark_ml/environments/single_agent_runtime.py:66]: image

This is what happens in the `examples/tfa.py` script (No changes from our side) when using the `ContinuousHighwayBlueprint`. Setting `params["World"]["remove_agents_out_of_map"] = False` prevents the error.

Is this intended behaviour and shall we remove the check if the agent is valid?

Comment (patrickhart) [reply bark_ml/environments/single_agent_runtime.py:66]: The fix as it is currently implemented is not a viable option as this would only work for the discrete behavior model.

I would suggest to add a log message:
```python
if eval_id not in world.agents:
  logger.info(f"Eval Agent-ID {eval_id} does not exist in the current world.")
```

And the flag should be additionally be set to true.

Comment (Brucknem) [reply bark_ml/environments/single_agent_runtime.py]: I removed the lines completely, as when we use the `params["World"]["remove_agents_out_of_map"] = False` flag the agent is always valid and the logging is never triggered. If we use `params["World"]["remove_agents_out_of_map"] = True` the logging is triggered, but the next line of code triggeres the IndexError and stops the execution.
